utils/src/main/java/software/amazon/awssdk/utils/cache/lru/LruCache.java

```diff
@@ -0,0 +1,236 @@
+import java.util.function.Function;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * An LRU (Least Recently Used) cache implementation.
+ * <p>
+ * The user can configure the maximum size of the cache, which is set to a default of 100.
+ */
+@SdkProtectedApi
+public final class LruCache<K, V>  {
+
+    private static final int DEFAULT_SIZE = 100;
+
+    private final Map<K, CacheEntry<K, V>> cache;
+    private final Function<K, V> valueSupplier;
+
+    private CacheEntry<K, V> leastRecentlyUsed = null;
+    private CacheEntry<K, V> mostRecentlyUsed = null;
+    private final Lock listLock = new ReentrantLock();
+    private final int maxCacheSize;
+
+    private LruCache(Builder<K, V> b) {
+        this.valueSupplier = b.supplier;
+        Integer customSize = Validate.isPositiveOrNull(b.maxSize, "size");
+        this.maxCacheSize = customSize != null ? customSize : DEFAULT_SIZE;
+        this.cache = new ConcurrentHashMap<>();
+    }
+
+    public V get(K key) {
+        CacheEntry<K, V> cachedEntry = cache.get(key);
+        if (cachedEntry != null) {
+            moveToFront(cachedEntry);
+            return cachedEntry.value();
+        }
+        V value = valueSupplier.apply(key);
+        CacheEntry<K, V> newEntry = new CacheEntry<>(key, value);
+        addToFront(newEntry);
+        return newEntry.value();
+    }
+
+    /**
+     * Sets an existing entry as the most recently used.
+     *
+     * 1) Remove the entry from its current place in the double linked list by letting its previous neighbor point to its
+     *    next neighbor, and vice versa.
+     * 2) This is the new most recently used item so there is no previous entry. Point to the most recently used entry.
+     * 3) Set the most recently used pointer to point to the new entry.
+     *
+     */
+    private void moveToFront(CacheEntry<K, V> entry) {
+        if (entry.equals(mostRecentlyUsed)) {
+            return;
+        }
+
+        boolean lockAcquired = listLock.tryLock();
+
+        try {
+            remove(entry);
+            entry.setPrevious(null);
+            entry.setNext(mostRecentlyUsed);
+            mostRecentlyUsed.setPrevious(entry);
+            mostRecentlyUsed = entry;
+        } finally {
+            if (lockAcquired) {
+                listLock.unlock();
+            }
+        }
+    }
+
+    /**
+     * Removes an entry from the cache. Remove the entry from its current place in the double linked list
+     * by letting its previous neighbor point to its next neighbor, and vice versa, if those exist.
+     */
+    private void remove(CacheEntry<K, V> entry) {
+        CacheEntry<K, V> previousEntry = entry.previous();
+        if (previousEntry != null) {
+            previousEntry.setNext(entry.next());
+        }
+        CacheEntry<K, V> nextEntry = entry.next();
+        if (nextEntry != null) {
+            nextEntry.setPrevious(entry.previous());
+        }
```

Comment:
if previous or next are null, does that mean we should update leastRecentlyUsed or mostRecentlyUsed (because they're probably pointing to this entry)?

Reply:
Yes, there seems to be a bug in there.